compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+
+    forwarderDef.withSpan(nextSymbol.span.shift(1))
+  }
+
+  def generateFromProduct(startParamIndices: List[Int], paramCount: Int, defdef: DefDef)(using Context) = {
+    cpy.DefDef(defdef)(
+      name = defdef.name,
+      paramss = defdef.paramss,
+      tpt = defdef.tpt,
+      rhs = Match(
+        ref(defdef.paramss.head.head.asInstanceOf[ValDef].symbol).select(termName("productArity")),
+        startParamIndices.map { paramIndex =>
+          val Apply(select, args) = defdef.rhs: @unchecked
+          CaseDef(
+            Literal(Constant(paramIndex)),
+            EmptyTree,
+            Apply(
+              select,
+              args.take(paramIndex) ++
+                Range(paramIndex, paramCount).map(n =>
+                  ref(defdef.symbol.owner.companionModule)
+                    .select(DefaultGetterName(defdef.symbol.owner.primaryConstructor.name.toTermName, n))
+                )
+            )
+          )
+        } ++ Seq(
+          CaseDef(
+            Underscore(defn.IntType),
+            EmptyTree,
+            defdef.rhs
+          )
+        )
+      )
+    ).setDefTree
+  }
+
+  def generateSyntheticDefs(tree: Tree, compute: ComputeIndicies)(using Context): Option[(Symbol, Option[Symbol], Seq[DefDef])] = tree match {
+    case defdef: DefDef if defdef.paramss.nonEmpty =>
+      import dotty.tools.dotc.core.NameOps.isConstructorName
+
+      val isCaseCopy =
+        defdef.name.toString == "copy" && defdef.symbol.owner.is(CaseClass)
+
+      val isCaseApply =
+        defdef.name.toString == "apply" && defdef.symbol.owner.companionClass.is(CaseClass)
+
+      val isCaseFromProduct = defdef.name.toString == "fromProduct" && defdef.symbol.owner.companionClass.is(CaseClass)
+
+      val annotated =
+        if (isCaseCopy) defdef.symbol.owner.primaryConstructor
+        else if (isCaseApply) defdef.symbol.owner.companionClass.primaryConstructor
+        else if (isCaseFromProduct) defdef.symbol.owner.companionClass.primaryConstructor
+        else defdef.symbol
+
+      compute(annotated) match {
+        case Nil => None
+        case Seq((paramClauseIndex, annotationIndices)) =>
+          val paramCount = annotated.paramSymss(paramClauseIndex).size
+          if isCaseFromProduct then
+            Some((defdef.symbol, Some(defdef.symbol), Seq(generateFromProduct(annotationIndices, paramCount, defdef))))
+          else
+            val (generatedDefs, _) =
+              val indices = (annotationIndices :+ paramCount).sliding(2).toList.reverse
+              indices.foldLeft((Seq.empty[DefDef], defdef.symbol)):
+                case ((defdefs, nextSymbol), Seq(paramIndex, nextParamIndex)) =>
+                  val forwarder = generateSingleForwarder(
+                    defdef,
+                    defdef.symbol.info,
+                    paramIndex,
+                    paramCount,
+                    nextParamIndex,
+                    nextSymbol,
```

Comment:
Is there a good reason that we don't make all the forwarders directly call `defdef.symbol`? That should be simpler and potentially more efficient, wouldn't it?

Reply:
I'd agree, this was a leftover from the mode that supported abstract methods

Reply:
the argument made was byte code reduction, its also in the SIP text if that matters, or can that be regarded as implementation detail?

Reply:
Either way works, one has a larger bytecode size and one has a larger runtime overhead. I chose the larger runtime overhead option, but semantically both should be identical

Reply:
yeah also it was relied upon by the abstract method stuff that is now removed

Reply:
actually, the implementation does call the def directly, nextSymbol is only being used to compute that shifted span offset - I'll try to avoid needing that